### DISCUSSION TITLE
How to add pores into a given network and assign them with different diameters?

### DISCUSSION THREAD

**Discussion by Camelot18** [PMEAL/porespy#624] [Q&A]
Dear Prof. Gostick, 

I want to add new pores into the created pore network to label them as "hydrate", assign them diameters, and put them in specific positions before exporting the data. However, the _openpnm.topotools.extend(network, coords=[], conns=[], labels=[], **kwargs)_ I used didn't work for adding new geometry size of hydrate(pores). Could you please help with this?

``` python
for i in range(np.size(x)):
    op.topotools.extend(network = net, coords=[(x[i],y[i],z[i])], labels=['hydrate'])
   
for j in net.filter_by_label(pores=net['pore.all'], labels='pore.hydrate'):
    net['pore.diameter'][j] = r_h3[1]*net['pore.diameter'][j-2954]
```

The returned length of net['pore.diameter'] is still 2954. How can I build the new pore network(the initial network + new pores) in this case?

Many thanks in advance!
Peter

**Comment by jgostick** [comment]
I'm not totally sure what your doing in the second loop, but the first one looks ok to me.  So you're saying that ``net.Np`` is 2954 both before AND after running that first loop?  Not that you could try to just add all the x,y,z coords at once.  I think 
``op.topotools.extend(network = net, coords=np.vstack((x, y, z)), labels=['hydrate'])`` should work (or use np.hstack, depending on the shape of x, y and z).

**Reply by Camelot18** [reply]
Sorry for the late reply.
The first loop will double the length of net['pore'](the initial size is 2954 and now turns to 5908), while the length of net['pore.hydrate'] is also 5908 but the first 2954 elements are 'false'. The second loop attempts to add geometry to the net['pore.hydrate'] but doesn't work. The idea is to creat the same number of spheres within the pores with customsized diameters and random positions, but I need to label them as 'hydrate'.

Besides, when I tried to use _op.io.STL.export_data(network=net, filename='imported_pn_hfs')_, it returned with "local variable 'csg' referenced before assignment". How to fix this?

Thanks again!
